### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by Rylan12** [Homebrew/homebrew-json#11]
## Overview of the current state

This tap is now in a good initial state. Currently, when `HOMEBREW_JSON_CORE` is set, homebrew/core can be untapped and `brew json` will install a formula by name using the bottle JSON API. Additionally, most `brew` commands work with homebrew/core untapped and on formulae installed using `brew json`.

## Known to-do items

This is a list of tasks that I know need to be completed before this becomes "officially usable":

- [ ] `brew update` should detect whether there are installed formulae that can be upgraded
- [ ] `brew upgrade` should work for formulae installed with `brew json`
- [ ] `brew reinstall` should work
- [ ] Restrict the use of `HOMEBREW_JSON_CORE` to users who haven't run a dev-cmd and haven't set `HOMEBREW_DEVELOPER`

## Future plan

Here is the current plan for the future (mostly based on @MikeMcQuaid's comment in https://github.com/Homebrew/brew/pull/11555#issuecomment-863980532:

1. Ask maintainers to play around with setting `HOMEBREW_JSON_CORE`, untapping homebrew/core, and installing formulae
2. Document the `HOMEBREW_JSON_CORE` variable and allow any user to use it
3. Make this functionality the default for new and existing Homebrew installs

## Questions

- How will users install a formula once homebrew/core is removed? Right now they need to use `brew json`. Will `brew install` be modified to support this (or, somehow, made into an alias for `brew json`?
- Will everyone who wants to have homebrew/core untapped need to have the homebrew/json tap installed?
- Should the contents of this tap be migrated to homebrew/brew? If so, at which stage in the above plan?
- Will new Homebrew installations auto-tap homebrew/core? What about homebrew/json?

**Comment by MikeMcQuaid** [comment]
> * `brew update` should detect whether there are installed formulae that can be upgraded

Can you elaborate a bit on this? I'm not sure I understand.



> * `brew upgrade` should work for formulae installed with `brew json`

Does this mean "they are upgraded as normal" or "they can also be upgraded without homebrew/core tapped"? The second seems desirable.

> 1. Ask maintainers to play around with setting `HOMEBREW_JSON_CORE`, untapping homebrew/core, and installing formulae

👍🏻 


> Right now they need to use `brew json`. Will `brew install` be modified to support this (or, somehow, made into an alias for `brew json`?

Yes, functionality should be moved into Homebrew/brew so that this tap is no longer required and `brew install` works  transparently. It may be worth handling `upgrade`/`reinstall` before then, up to you.

> * Will everyone who wants to have homebrew/core untapped need to have the homebrew/json tap installed?

For now: yes. I expect once we get to "Ask maintainers to play around with setting `HOMEBREW_JSON_CORE`" and beyond: this tap can get archived/deleted as all functionality will be in Homebrew/brew.

> * Should the contents of this tap be migrated to homebrew/brew? If so, at which stage in the above plan?

Yes but merged into the existing `install` commands/logic (hopefully with some code separation). I would say either now or after `upgrade` and `reinstall` are working.

> * Will new Homebrew installations auto-tap homebrew/core? 

Yes, for now. Once we are only auto-tapping on dev-cmd (or maybe even failing on dev-cmd unless the user manually taps): I think we should stop auto-tapping during installation.

> What about homebrew/json?

No.

**Comment by Rylan12** [comment]
> > * `brew update` should detect whether there are installed formulae that can be upgraded
> 
> Can you elaborate a bit on this? I'm not sure I understand.

Yeah, good question. It's possible that `brew update` will just ignore the fact that homebrew/core isn't tapped (that's what's happening now), but I was thinking that it would check for outdated installed formulae. In my Homebrew workflow, I don't run `brew upgrade` unless I've just run `brew update` and it tells me that there are outdated formulae/casks. So, for a seamless experience with having homebrew/core untapped, I would want `brew update` to check all my installed formulae to see whether their versions match the latest version from the bottle API. If there's an upgrade available, display that information so I know to next run `brew upgrade`.

> > * `brew upgrade` should work for formulae installed with `brew json`
> 
> Does this mean "they are upgraded as normal" or "they can also be upgraded without homebrew/core tapped"? The second seems desirable.

I'm not totally sure what you mean by "upgraded as normal" but what I mean here is that you can run `brew upgrade` on an installed formula when homebrew/core is not tapped and it will check whether there's a new version available and upgrade. Right now, if you run `brew upgrade`, it just looks at the only copy of the formula it can find (in the keg), sees that the version matches, and says that the latest version is already installed.

> > Right now they need to use `brew json`. Will `brew install` be modified to support this (or, somehow, made into an alias for `brew json`?
> 
> Yes, functionality should be moved into Homebrew/brew so that this tap is no longer required and `brew install` works transparently. It may be worth handling `upgrade`/`reinstall` before then, up to you.

> Yes but merged into the existing `install` commands/logic (hopefully with some code separation). I would say either now or after `upgrade` and `reinstall` are working.

👍 So it sounds like this (and maybe the `upgrade`/`reinstall` stuff) should be done before asking maintainers to test this out?

> > * Will new Homebrew installations auto-tap homebrew/core?
> 
> Yes, for now. Once we are only auto-tapping on dev-cmd (or maybe even failing on dev-cmd unless the user manually taps): I think we should stop auto-tapping during installation.

👍

**Comment by MikeMcQuaid** [comment]
> Yeah, good question. It's possible that `brew update` will just ignore the fact that homebrew/core isn't tapped (that's what's happening now), but I was thinking that it would check for outdated installed formulae. In my Homebrew workflow, I don't run `brew upgrade` unless I've just run `brew update` and it tells me that there are outdated formulae/casks. So, for a seamless experience with having homebrew/core untapped, I would want `brew update` to check all my installed formulae to see whether their versions match the latest version from the bottle API. If there's an upgrade available, display that information so I know to next run `brew upgrade`.

Gotcha. `brew update` would still have to be run in some form to keep e.g. Homebrew/brew (and homebrew-cask although that's another super slow repository so might be worth thinking how we can move towards that being untapped eventually too...) up-to-date. I'd suggest that it call into some code shared with e.g. `brew outdated` to be able to identify what's outdated purely based on the formulae.brew.sh JSON.

> Edit: ah, I forgot about `brew outdated`. What I'm describing above should definitely go in that command. I think I would still like it to be in `brew update` because I don't really use `brew outdated` in my workflow, but if we think it doesn't really make 

:+1:

> I'm not totally sure what you mean by "upgraded as normal" but what I mean here is that you can run `brew upgrade` on an installed formula when homebrew/core is not tapped and it will check whether there's a new version available and upgrade. Right now, if you run `brew upgrade`, it just looks at the only copy of the formula it can find (in the keg), sees that the version matches, and says that the latest version is already installed.

Makes sense. I guess it may be worth thinking what the behaviour should be if `HOMEBREW_NO_AUTO_UPDATE` is disabled. It may make sense to just refuse to use the JSON path if that's set because we even moreso cannot stick on an older version.



> 👍 So it sounds like this (and maybe the `upgrade`/`reinstall` stuff) should be done before asking maintainers to test this out?

Think so :+1:

--- 

Great work on all this @Rylan12, you're killing it!
